bt: services: gap_default: Extract GAP service from gatt implementation

rkel opened the pull request:
Remove Generic Access Service implementation from gatt files and move it into separate file in services directory.
The proposed name is just gap_default as it is default implementation that may be switched off by the user in purpose of providing the GAP implementation directly in the application.

In the end it serves the same purpose that #93946 but gives the user total freedom of implementation when needed in the same time being 100% compatible with the current solution.

I was testing it manually and now I can enable and disable default GAP implementation.

The main reason of this PR is to discuss if we wish to go into this solution instead of #93946.

To be done:
[v] - GAP service validation during bt_enable.
[v] - review all configurations that need to be moved to Kconfig.gap_default
[v] - test of the functionality (BabbleSim)
[v] - sample with GAP reimplementation on App side
[v] - update documentation

rkel committed:
Bluetooth: Host: Extract GAP service from GATT

Remove Generic Access Service implementation from gatt files and move it
into separate file in services directory.
The default service implementation can be disabled by the user in purpose
of providing the GAP implementation directly in the application.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>

alwa-nordic commented on subsys/bluetooth/services/Kconfig.gap_default:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2025 Koppel Electronic
+# SPDX-License-Identifier: Apache-2.0
+
+config BT_GAP_DEFAULT
```
This config name is too generic. The GAP Service is only one component of GAP. I suggest something like `BT_GAP_SVC_DEFAULT_IMPL`.

alwa-nordic commented on subsys/bluetooth/services/gap_default.c:
```diff
@@ -0,0 +1,188 @@
+				 strlen(name));
+}
+
+#if defined(CONFIG_BT_DEVICE_NAME_GATT_WRITABLE)
```
The configs for the GAP Service should be moved `Kconfig.gap_default`.

jhedberg commented on subsys/bluetooth/services/gap_default.c:
```diff
@@ -0,0 +1,188 @@
+#endif
+
+#if defined (CONFIG_BT_GAP_PERIPHERAL_PREF_PARAMS)
+/* This checks if the range entered is valid */
+BUILD_ASSERT(!(CONFIG_BT_PERIPHERAL_PREF_MIN_INT > 3200 &&
+	     CONFIG_BT_PERIPHERAL_PREF_MIN_INT < 0xffff));
+BUILD_ASSERT(!(CONFIG_BT_PERIPHERAL_PREF_MAX_INT > 3200 &&
+	     CONFIG_BT_PERIPHERAL_PREF_MAX_INT < 0xffff));
+BUILD_ASSERT(!(CONFIG_BT_PERIPHERAL_PREF_TIMEOUT > 3200 &&
+	     CONFIG_BT_PERIPHERAL_PREF_TIMEOUT < 0xffff));
+BUILD_ASSERT((CONFIG_BT_PERIPHERAL_PREF_MIN_INT == 0xffff) ||
+	     (CONFIG_BT_PERIPHERAL_PREF_MIN_INT <=
+	     CONFIG_BT_PERIPHERAL_PREF_MAX_INT));
+BUILD_ASSERT((CONFIG_BT_PERIPHERAL_PREF_TIMEOUT * 4U) >
+	     ((1U + CONFIG_BT_PERIPHERAL_PREF_LATENCY) *
+	      CONFIG_BT_PERIPHERAL_PREF_MAX_INT));
```
Yes, seems like they should move to some GAP-specific place. Btw, some of those build asserts should probably be converted into `range` directives in Kconfig.

rkel replied:
@jhedberg @alwa-nordic - should I just leave this assertions in the gatt.c for the purpose of this PR?

rkel replied:
I will do it right now - it looks little odd as now we have only the parameter checking in the gatt.c. But I get it that it is better not to move it to the module that may be disabled.

rkel committed:
Bluetooth: gap_svc_default: Set user_data for name characteristic to NULL

Set the unused argument of the name characteristic to NULL.
This removes the reference from GAP service to the hci_core.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>

rkel committed:
Bluetooth: Host: Conn: Add GAP service validation

Implement the validation if there is exactly one GAP service
in GATT database.
The validation is performed as soon as the GATT database
is available and forbids to enable the Bluetooth if the configuration
is not compatible with the Bluetooth Specification.
The verification may be disabled by the user for better performance
or deliberate implementation that is not compliant with the
specification.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>

rkel committed:
tests: bluetooth: host: conn: Disable GATT GAP svc validation

Disable the GATT GAP service validation for the test.

This test mocks bluetooth modules beside of conn and net_buf.
The mock does not contain functionality required by the validation module.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>

rkel committed:
samples: bluetooth: peripheral_gap_svc: Adding sample

Adding a sample that presents how to implement non default GAP service
inside the application.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>

rkel commented on samples/bluetooth/peripheral_gap_svc/README.rst:
```diff
@@ -0,0 +1,29 @@
+This sample demonstrates the usage of the NUS service (Nordic UART Service) as a serial
+endpoint to exchange data. In this case, the sample assumes the data is UTF-8 encoded,
+but it may be binary data. Once the user connects to the device and subscribes to the TX
+characteristic, it will start receiving periodic notifications with "Hello World!\n".
```
I just did not saved the changes - will update it in nearest push.

rkel commented on samples/bluetooth/peripheral_gap_svc/README.rst:
```diff
@@ -0,0 +1,29 @@
+Overview
+********
+
+This sample demonstrates the implementation of the GAP service
```
Fix in next push

rkel commented on subsys/bluetooth/services/Kconfig.gap_svc:
```diff
@@ -0,0 +1,76 @@
+config BT_DEVICE_APPEARANCE_GATT_WRITABLE_NONE
+	bool "No requirements"
+
+config BT_DEVICE_APPEARANCE_GATT_WRITABLE_ENCRYPT
```
This is direct copy of the existing settings name. We can change it later if we decide it makes sense, but lets adjust it in other PR - this would probably require to mark old setting obsolete to give the user a time to migrate.

rkel commented on subsys/bluetooth/services/Kconfig.gap_svc:
```diff
@@ -0,0 +1,76 @@
+config BT_DEVICE_APPEARANCE_GATT_WRITABLE_NONE
+	bool "No requirements"
+
+config BT_DEVICE_APPEARANCE_GATT_WRITABLE_ENCRYPT
```
This is direct copy of the existing settings name. We can change it later if we decide it makes sense, but lets adjust it in other PR - this would probably require to mark old setting obsolete to give the user a time to migrate.

Thalley commented on include/zephyr/bluetooth/gap.h:
```diff
@@ -26,6 +26,15 @@ extern "C" {
+ * GAP service implementation.
+ * This ensures that the GAP service would be placed just after BT_UUID_GATT.
+ */
+#define BT_GATT_GAP_SVC_DEFAULT_NAME _2_gap_svc
```
This should either be defined in `gatt.h`, or renamed to 
```suggestion
#define BT_GAP_GATT_SVC_DEFAULT_NAME _2_gap_svc
```

Is it a requirement that GAP service is placed there in the database? If not, should this be a Kconfig option?

rkel replied:
This was an approach that was here to make sure that GAP service is placed just after GATT service (named _1_gatt_svc. But to be honest - it seems not to be a requirement according to specification. It is rather a convention. But as far as I know, the specification only requires GAP to be present not adding any more restrictions about its placement in the database. We can probably just remove it.

rkel replied:
I will just remove it right now - but I had also placed a question at bt channel to be double sure. If I am now right here - I will get this definition back renaming it properly.

Thalley commented on samples/bluetooth/peripheral_gap_svc/src/main.c:
```diff
@@ -0,0 +1,157 @@
+	const char *name = bt_get_name();
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, name,
+				 strlen(name));
```
Minor: We use a max width of 100 characters in Zephyr, so this could be


```suggestion
	return bt_gatt_attr_read(conn, attr, buf, len, offset, name, strlen(name));
```

Thalley commented on samples/bluetooth/peripheral_gap_svc/src/main.c:
```diff
@@ -0,0 +1,157 @@
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LEN),
```
If `CONFIG_BT_DEVICE_NAME_GATT_WRITABLE=y` then we shouldn't use `CONFIG_BT_DEVICE_NAME` for the advertising data, as if it's changed then we are advertising an incorrect name

Thalley commented on subsys/bluetooth/host/gatt_gap_svc_validate.c:
```diff
@@ -0,0 +1,55 @@
+				    uint16_t handle,
+				    void *user_data)
+{
+	size_t *cnt = user_data;
```
Consider if `cnt` should be changed to `bool found`; do we care about how many services we find, or do we simply care that we found one and only one?

Thalley commented on subsys/bluetooth/host/gatt_gap_svc_validate.h:
```diff
@@ -0,0 +1,27 @@
+ * @return 0 on success (exactly one GAP service found).
+ * @return -EINVAL if zero or multiple GAP services are found.
+ */
+int gatt_gap_svc_validate(void);
```
We could also consider making this a `bool` return type

Thalley commented on subsys/bluetooth/services/gap_svc_default.c:
I don't want to review and apply comments to code in this file that you didn't change. Was any changes made, or was it just copied as-is from gatt.c?

Thalley commented on tests/bsim/bluetooth/host/gatt/gap_svc/src/central.c:
```diff
@@ -0,0 +1,359 @@
+
+			memcpy(&u16, &data->data[i], sizeof(u16));
+			uuid = BT_UUID_DECLARE_16(sys_le16_to_cpu(u16));
+			if (bt_uuid_cmp(uuid, BT_UUID_HRS)) {
```
```suggestion
			if (bt_uuid_cmp(uuid, BT_UUID_HRS) != 0) {
```

Thalley commented on tests/bsim/bluetooth/host/gatt/gap_svc/src/central.c:
```diff
@@ -0,0 +1,359 @@
+				create_param->options &= ~BT_CONN_LE_OPT_CODED;
+				err = bt_conn_le_create(addr, create_param,
+							param, &default_conn);
+				if (err) {
```
```suggestion
				if (err != 0) {
```

Thalley commented on tests/bsim/bluetooth/host/gatt/gap_svc/src/central.c:
```diff
@@ -0,0 +1,359 @@
+	 */
+	struct bt_le_scan_param scan_param = {
+		.type       = BT_LE_SCAN_TYPE_ACTIVE,
+		.options    = BT_LE_SCAN_OPT_CODED,
```
Ditto; just scan without codec PHY for simplicity

Thalley commented on tests/bsim/bluetooth/host/gatt/gap_svc/src/central.c:
```diff
@@ -0,0 +1,359 @@
+		printk("Scanning without Coded PHY\n");
+		scan_param.options &= ~BT_LE_SCAN_OPT_CODED;
+		err = bt_le_scan_start(&scan_param, device_found);
+		if (err) {
```
```suggestion
		if (err != 0) {
```

Thalley reviewed (changes requested):
First round of review :) 

Firstly, thanks for the contribution. Overall looks good. Most of my comments are related to coding style and coding guidelines violations, which probably mostly comes from you copying existing code that are older than these guidelines, and thus does not follow them (but new code should).

Thalley commented on tests/bsim/bluetooth/host/gatt/gap_svc/src/peripheral.c:
Since you are adding a peripheral sample for a custom GAP service, it is actually possible to use that sample as the peripheral in your test. That way you can significantly reduce code duplication, while also actively testing the sample. See e.g. https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bsim/bluetooth/samples for inspiration

Thalley replied:
I'll wait with reviewing this file more thoroughly since most of my comments would be the same as the ones I posted in the sample itself

rkel replied:
Done

rkel commented:
^^ Just a rebase before any work.

rkel commented on subsys/bluetooth/services/gap_svc_default.c:
```diff
@@ -0,0 +1,175 @@
+#if defined(CONFIG_DEVICE_APPEARANCE_GATT_WRITABLE_AUTHEN)
+	#define GAP_APPEARANCE_PERMS (BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_AUTHEN)
+#elif defined(CONFIG_BT_DEVICE_APPEARANCE_GATT_WRITABLE_ENCRYPT)
+	#define GAP_APPEARANCE_PERMS (BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT)
```
No - it should not.

rkel commented on tests/bsim/bluetooth/host/gatt/gap_svc/src/peripheral.c:
```diff
@@ -0,0 +1,256 @@
+	/* adding one to fit the terminating null character */
+	char value[CONFIG_BT_DEVICE_NAME_MAX + 1] = {};
+
+	if (offset != 0) {
```
Fixed

rkel commented on subsys/bluetooth/services/gap_svc_default.c:
```diff
@@ -0,0 +1,175 @@
+
+#if defined(CONFIG_BT_DEVICE_APPEARANCE_GATT_WRITABLE)
+	#define GAP_APPEARANCE_PROPS (BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE)
+#if defined(CONFIG_DEVICE_APPEARANCE_GATT_WRITABLE_AUTHEN)
```
It still is just a copy of old settings. Keeping compatibility.

rkel commented on subsys/bluetooth/services/Kconfig.gap_svc:
```diff
@@ -0,0 +1,76 @@
+	bool "Encryption required"
+
+config DEVICE_NAME_GATT_WRITABLE_AUTHEN
+	bool "Encryption and authentication required"
```
Compatibility.

rkel commented on tests/bluetooth/host/conn/prj.conf:
```diff
@@ -9,6 +9,7 @@ CONFIG_BT_CENTRAL=y
 CONFIG_BT_EXT_ADV=y
 CONFIG_BT_PER_ADV=y
 CONFIG_BT_PER_ADV_SYNC=y
+CONFIG_GATT_GAP_SVC_VALIDATE=n
```
OK

rkel commented on samples/bluetooth/peripheral_gap_svc/README.rst:
```diff
@@ -0,0 +1,25 @@
+In this case the GAP name write operation is limited to always start with capital letter.
+Once the user connects to the device and tries to set its name, only names starting
+with capital letter would be accepted.
+Additionally the fact that the name has been change would be logged immediately.
```
Implemented

rkel commented on samples/bluetooth/peripheral_gap_svc/src/main.c:
```diff
@@ -0,0 +1,157 @@
+	if (value[0] < 'A' || value[0] > 'Z') {
+		printk("Rejected name change to \"%s\": must start with capital letter\n", value);
+		return BT_GATT_ERR(BT_ATT_ERR_VALUE_NOT_ALLOWED);
+	}
```
No - this is the functionality of the sample. Just a demonstration what can it be used for. It is described in the sample README.rst:
```This sample demonstrates the implementation of the GAP service on the application side.
In this case the GAP name write operation is limited to always start with capital letter.
Once the user connects to the device and tries to set its name, only names starting
with capital letter would be accepted.
```

rkel commented on samples/bluetooth/peripheral_gap_svc/src/main.c:
```diff
@@ -0,0 +1,157 @@
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
```
Fixed to: if (conn_err != BT_HCI_ERR_SUCCESS)

rkel commented on samples/bluetooth/peripheral_gap_svc/src/main.c:
```diff
@@ -0,0 +1,157 @@
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
```
Well.... it is event not possible in theory:
1. From documentation of bt_addr_le_to_str it returns number of successfully formatted bytes from binary address.
2. From implementation: It uses snprintk, that would return error only if there is formatting string error - that would not happen as it is constant value here and user cannot change it.

Theoretically we can get less characters than the buffer size, but snprintk makes sure that the string is always NULL terminated, so only situation that we may have here is not printing the whole address in logs. Moreover we are using the buffer size of BT_ADDR_LE_STR_LEN - so I am not sure if we wish to learn the users to check the value returned here and process it. We can argue if it would not break rule 19 (about unreachable code).

Just looking through the code I can see not a sample that tests the value returned here. The only thing that I can see is that come samples uses a form (void)bt_addr_le_to_str(...), but they are in minority.

rkel replied:
@Thalley - please note also that you have provided documentation for snprintf, while ```bt_addr_le_to_str``` uses snprintk and this is the one we have total control of. It uses cbvprintf, using str_out as the way to put the string characters into the buffer. The str_out really never returns negative values into ctx->count. And ctx->count is the value we are returning from snprintk finally. I had analyzed it twice: manually, and then asked Claude Sonnet 4.5 if it can return negative value. The code is really simple, and easy to follow in printk.c file - but just to be sure that I am not tired and cannot see something obvious. The answer was twice: There is no possiblity that vsnprintk return negative value.

And it means also I was tried last time pointing out that snprintk may return negative value if there is an error in formatting string :)

rkel commented on subsys/bluetooth/host/gatt_gap_svc_validate.c:
```diff
@@ -0,0 +1,55 @@
+	size_t *cnt = user_data;
+	const struct bt_uuid *svc_uuid = attr->user_data;
+
+	if (svc_uuid && bt_uuid_cmp(svc_uuid, BT_UUID_GAP) == 0) {
```
Fixed

rkel commented on subsys/bluetooth/host/gatt_gap_svc_validate.c:
```diff
@@ -0,0 +1,55 @@
+		if (*cnt > 1) {
+			LOG_ERR("Multiple (%zu) GAP services found at handle: %u",
+				*cnt, handle);
+		}
```
After longer consideration - I just had removed the whole if (*cnt > 1) condition here.

rkel commented on subsys/bluetooth/host/conn.c:
```diff
@@ -4428,6 +4429,14 @@ int bt_conn_init(void)
+	if (IS_ENABLED(CONFIG_GATT_GAP_SVC_VALIDATE)) {
+		err = gatt_gap_svc_validate();
+		if (err) {
+			LOG_ERR("GATT GAP service validation failed (err %d)", err);
```
After longer consideration - I had removed the error logging from gap_service_check_cb and had changed the log in gatt_gap_svc_validate from error to debug level.

rkel commented on subsys/bluetooth/host/gatt_gap_svc_validate.c:
```diff
@@ -0,0 +1,55 @@
+
+int gatt_gap_svc_validate(void)
+{
+	size_t gap_svc_count = 0;
```
This is very good rule. I usually add it as I hit nasty error forgetting it once when shifting the value left :). Fixed.

rkel committed:
Bluetooth: gap_svc_default: Update attribute len checking in write_name

Remove offset form the attribute length checking as the offset
is tested before to be 0.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>

rkel commented on tests/bsim/bluetooth/host/gatt/gap_svc/src/peripheral.c:
```diff
@@ -0,0 +1,256 @@
+/* Name changed called in the test */
+bool gap_svc_name_changed;
+/* Appearance changed called in the test */
+bool gap_svc_appearance_changed;
```
Fixed.

rkel committed:
tests: bsim: Bluetooth: peripheral_gap_svc: Implementation

Implementation of the bsim test for peripheral_gap_svc sample.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>

rkel commented on samples/bluetooth/peripheral_gap_svc/src/main.c:
```diff
@@ -0,0 +1,159 @@
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/bluetooth/gatt.h>
+
```
OK

jhedberg reviewed (dismissed):
This looks ok, except that the incorrect `CONFIG_DEVICE_*` prefixes should be fixed to `CONFIG_BT_DEVICE_*` along with the appropriate deprecation process. I'll leave it to the author to decide whether to do that here or in a subsequent PR.

rkel committed:
Bluetooth: gap_svc_default: Add BT_ prefix to DEVICE_xxx_GATT_WRITABLE_

Rename the DEVICE_NAME_GATT_WRITABLE_ kconfig options to
BT_DEVICE_NAME_GATT_WRITABLE_.
Rename DEVICE_APPEARANCE_GATT_WRITABLE_AUTHEN to
BT_DEVICE_APPEARANCE_GATT_WRITABLE_AUTHEN.

The old options are marked as deprecated.

Signed-off-by: Radosław Koppel <r.koppel@k-el.com>

rkel commented:
@jhedberg > This looks ok, except that the incorrect `CONFIG_DEVICE_*` prefixes should be fixed to `CONFIG_BT_DEVICE_*` along with the appropriate deprecation process. I'll leave it to the author to decide whether to do that here or in a subsequent PR.

Names changed. I am only worrying about the fact that BT_DEVICE_NAME_GATT_WRITABLE and BT_DEVICE_NAME_GATT_WRITABLE are both used as boolean selector and choice name. Should we change it too?

rkel commented on subsys/bluetooth/services/Kconfig.gap_svc:
```diff
@@ -0,0 +1,76 @@
+	prompt "Security requirements"
+	default DEVICE_APPEARANCE_GATT_WRITABLE_AUTHEN
+	help
+	  Select security requirements for writing device appearance by remote
```
@jhedberg @Thalley Is it ok or should I change it too?

Thalley replied:
I agree, that there is a bug here - I'm just suggesting fixing that in a separate PR (or at least commit). I'm not sure what the right way forward is when there's a clear but in an "public API" like this - @jhedberg thoughts?

Thalley commented on subsys/bluetooth/services/Kconfig.gap_svc:
```diff
@@ -0,0 +1,98 @@
+menu "GATT Generic Access Service"
+
+config BT_GAP_SVC_DEFAULT_IMPL
+	bool "GATT Generic Access Service - default implementation"
```
Not sure if my last comment w.r.t. the name here was considered, but should we really append `_DEFAULT_IMPL` here? 

We don't do that for any other GATT services that we have in the stack, even when they all act as default service implementations that any user can replace with their own implementations like this.

rkel replied:
Makes sense. I thing I did not get you first. I am not sure but from the top of my mind the name comes from previous discussions, but probably I started with gap_default, and then after the descussion it was changed to gap_svc_default. It makes sense to rename it to be just a service. I will take care of this now.

rkel replied:
Quickly checking for me it seems that we have good separation between service related settings and the settings related to GAP implementation in Zephyr. To be 100% sure, I will just try to change the scheme and check if anything breaks.

Thalley reviewed (changes requested):
Small handful of minor mostly minor comments, otherwise LGTM

rkel commented on subsys/bluetooth/host/Kconfig.gatt_gap_svc_validate:
I can see it now. The only exception is in Bluetooth audio. It makes it little harder to manage, but makes sense from the user perspective.

rkel replied:
Done

Thalley reviewed (approved):
One minor coding style violation still remains, but I don't want to block this entire PR on that :D
